WAFs might reject requests permanently (#3145)
httpwg/http-extensions: draft-ietf-httpbis-incremental.md

# Security Considerations

When receiving an incremental request, intermediaries might reject the request
due to security concerns. The following subsections explore typical scenarios

> under which the intermediaries might reject the requests.

Reviewer: ```suggestion
under which the intermediaries might reject requests.
```

## Permanent Rejection

> Some intermediaries inspect the payload of an HTTP request and forward them only if

Reviewer: ```suggestion
Some intermediaries inspect the payload of an HTTP messages and forward them only if
```

> the request is deemed safe. Such intermediaries might decide to reject

Reviewer: ```suggestion
the content is deemed safe. Such intermediaries might decide to reject
```

(though see below)

> the request is deemed safe. Such intermediaries might decide to reject
> incremental requests that do not carry sufficient information in its first
> flights of payload necessary to determine the safety of the request.

Reviewer: ```suggestion
the content is deemed safe. Any feature that depends on seeing the entirety
of the message in this way is incompatible with incremental delivery,
so these intermediaries need to reject requests unless the entire message
is received.
```

Author: Thanks, applied manually.

When an intermediary rejects an incremental request due to security concerns
with regard to the payload that the request might convey, the intermediary
SHOULD respond with a 403 Forbidden error with an http_request_denied Proxy-Status

> response header field ({{Section 2.3.17 of PROXY-STATUS}}).

Reviewer: There are two cases here:

1. incremental requests, which this text assumes and for which 403 etc is fine.
2. incremental *responses*, which I think also get some sort of rejection, but it won't be a http_request_denied, maybe http_response_timeout with 504.

The suggestions above try to make that text more generic, but I don't quite know how to make this text more general without spelling that out.

Author: +1 to making the problem statement generic, so I've applied your suggestions against the previous paragraph.

Regarding what the signal should be, I think I'm more inclined to defining the `incremental_unsupported` Proxy-Status error, as that can be the source of truth for the rejection of both requests and responses.

Author: Added as `incremental_refused` in commits up to 4b4bfc7.

# Security Considerations

When receiving an incremental request, intermediaries might reject the request
due to security concerns. The following subsections explore typical scenarios
under which the intermediaries might reject requests.

## Permanent Rejection

Some intermediaries inspect the payload of an HTTP messages and forward them
only if the content is deemed safe. Any feature that depends on seeing the
entirety of the message in this way is incompatible with incremental delivery,
so these intermediaries need to reject requests unless the entire message is
received.

When an intermediary rejects an incremental request due to security concerns
with regard to the payload that the request might convey, the intermediary
SHOULD respond with a 403 Forbidden error with an incremental_refused

> Proxy-Status response header field ({{iana-considerations}}).

Reviewer: Did you want to update this with the 504 response rejection case?

Reviewer 2: Please not 5xx. All infrastructure operators are monitoring those as a reliable metric for their infrastructure's health. 504 is a timeout, it happens when servers are too slow to respond (overloaded or depending on an overloaded database) and it's long been graphed and monitor as representing this. And ideally upon 504, a client should refrain from trying again too soon. Here the only reason for not processing is the presence of the incremental header in the request, i.e. if the client knows how to access the same feature without incremental, it will be immediately served.

Reviewer 2: Ah you meant about the response, I thought you were speaking about the request. Yes the response in this case must necessarily be a 5xx since we've blocked between the intermediary and the server, i.e. the client cannot fix this. It could be a problem by the way that we reject valid server responses in this case without being able to warn them that we can't proceed with incremental, because the user will face a dirty error while the application could have degraded much more nicely instead (e.g. using redirects to a better capable server, switching to a less interactive but compatible mechanism, such as a Refresh header etc). We leave it no chance with the 5xx like this.

Wouldn't it be good first if we could suggest that intermediaries that *have* to buffer responses send "Incremental: ?0" to the server to say "don't ever think about trying this". Because then it would make sense to consider that the server trying it anyway is violating the rule, and many more 5xx would be suitable here (including 502 Bad Gateway in this case).

Reviewer 2: I think I'm fine with 501, as it indeed matches an incompatibility between what the client desires and what the server is able to do, i.e. one of them has to change something for the request to be accepted. So yeah, that's fine for me.

I'm less sure that it's suitable for the response, because in order to block the response, the server needs to have already processed the request and started to respond. But 501 normally implies that the request was not processed. IMHO the 502 bad gateway is more suitable there as it's a deployment issue: the server is trying to do stuff that is incompatible with the hosting infrastructure. That's typically what's returned when a server response is invalid or blocked by an intermediary by policy (e.g. after detecting a risk of information leak etc).

Also I **do** think that we should support "Incremental: ?0" so that intermediaries can warn servers "don't try to do that, I'm not compatible" in order for servers to degrade nicely or return a clean message to the user instead of failing hard on a 502. And the 502 would be even more suitable there as it would be clear that the server didn't consider the advertised limitation.

# Security Considerations

When receiving an incremental request, intermediaries might reject the request
due to security concerns. The following subsections explore typical scenarios
under which the intermediaries might reject requests.

## Permanent Rejection

Some intermediaries inspect the payload of an HTTP messages and forward them
only if the content is deemed safe. Any feature that depends on seeing the
entirety of the message in this way is incompatible with incremental delivery,
so these intermediaries need to reject requests unless the entire message is
received.

When an intermediary rejects an incremental message -- either a request or a
response -- due to security concerns with regard to the payload that the message

> might convey, the intermediary SHOULD respond with a 501 Not Implemented error

Reviewer: ```suggestion
might convey, the intermediary SHOULD respond with a 501 (Not Implemented) error
```

: None
{: spacing="compact"}

An HTTP Proxy Error Type is registered in the HTTP Proxy Error Types registry as
shown below:

Name:
: incremental_refused

Description:
: The HTTP message contained the Incremental HTTP header field, but the
  intermediary refused to forward the message incrementally.

Extra Parameters:
: none

Recommended HTTP Status Code:

> : 403 if refusing to serve an incremental request, or 504 if refusing an
>   incremental response.

Reviewer: ```suggestion
: 501
```

Reviewer 2: Also I really think that if we offer the ability for the intermediary to warn the server about lack of incremental support, all of this becomes less of a problem since this case will then be marginal.
